google/cloud/aiplatform/prediction/handler.py

```diff
@@ -0,0 +1,106 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Optional, Type
+
+try:
+    from fastapi import Request
+    from fastapi import Response
+except ImportError:
+    raise ImportError(
+        "FastAPI is not installed and is required to build model servers. "
+        'Please install the SDK using "pip install python-aiplatform[prediction]"'
+    )
+
+from google.cloud.aiplatform.prediction.predictor import Predictor
+from google.cloud.aiplatform.prediction.serializer import DefaultSerializer
+
+
+class Handler:
+    """Interface for Handler class to handle prediction requests."""
+
+    def __init__(
+        self, gcs_artifacts_uri: str, predictor: Optional[Type[Predictor]] = None,
+    ):
+        """Initializes a Handler instance.
+
+        Args:
+            gcs_artifacts_uri (str):
+                Required. The value of the environment variable AIP_STORAGE_URI.
+            predictor (Type[Predictor]):
+                Optional. The Predictor class this handler uses to initiate predictor
+                instance if given.
+        """
+        pass
+
+    def handle(self, request: Request) -> Response:
+        """Handles a prediction request.
+
+        Args:
+            request (Request):
+                The request sent to the application.
+
+        Returns:
+            The response of the prediction request.
+        """
+        pass
+
+
+class PredictionHandler(Handler):
+    """Default prediction handler for the prediction requests sent to the application."""
+
+    def __init__(
+        self, gcs_artifacts_uri: str, predictor: Optional[Type[Predictor]] = None,
+    ):
+        """Initializes a Handler instance.
+
+        Args:
+            gcs_artifacts_uri (str):
+                Required. The value of the environment variable AIP_STORAGE_URI.
+            predictor (Type[Predictor]):
+                Optional. The Predictor class this handler uses to initiate predictor
+                instance if given.
+        """
+        if predictor is None:
+            raise ValueError(
+                "PredictionHandler must have a predictor class passed to the init function."
+            )
+
+        self._predictor = predictor()
+        self._predictor.load(gcs_artifacts_uri)
+
+    async def handle(self, request: Request) -> Response:
+        """Handles a prediction request.
+
+        Args:
+            request (Request):
+                Required. The prediction request sent to the application.
+
+        Returns:
+            The response of the prediction request.
+        """
+        request_body = await request.body()
+        prediction_input = DefaultSerializer.deserialize(
+            request_body, request.headers.get("content-type")
```

Comment:
Can we default this to json if the content-type is not specified?

And same with accept below.

Reply:
If the body is not in json format, then it will throw json deserialize error. Is it expected?

Reply:
I think it's fine. If the user doesn't provide a content-type, let's treat it as json by default.

Can we make the error message clear that the json deserialization failed? And also something like...

To specify a different type, please set the "content-type" header in the request. Currently supported content types: {...}.

Reply:
Updated as our offline discussion. PTAL.